Improve the PDF drill rendering

vdye opened the pull request:
This pull request has a lot going on, so apologies for that. It's essentially the beginning of an overhaul of field rendering to replace `FootballField.java` with a more generic `FieldView` class (inherited and modified as needed per use case).  Here, we make `PdfImage` inherit from `FieldView`, then adjust the rendering to be both higher resolution and better quality by enabling interpolation & anti-aliasing.

The pull request can be broken down as follows:
- Commit 1 scales `drillHeight` in the PDF including the sideline width.
- Commits 2-4 add helper methods for both `Point` & `RankPosition` that will be helpful in the overhauled rendering code.
- Commit 5 adds the `FieldStyle` class, which contains all of the information needed for styling the appearance of the football field (colors, line thicknesses, etc.)
- Commit 6 adds `FieldView`, which re-implements a lot of `FootballField`, but dynamically scaling by units of on-field feet rather than hardcoded pixel sizes. The drill generated is tweaked slightly to be more readable (e.g. the grid lines are less obtrusive, the endzone & sideline margins are explicitly included), but overall it still looks mostly the same as the `FootballField` rendering.
- Commit 7 adds a comment about a mysterious 0.6ft error in the rank positioning. Added an issue to address that long-term: https://github.com/bigredbands/rank-panda/issues/42
- Commit 8 enables anti-aliasing, bilinear interpolation, and generally "nicer" rendering for the field.
- Commit 9 increases the resolution of the printed drill image in the PDF to 300dpi, making it much easier to read when printed.

### Future work
Other things I'm planning to do that will build off of this:
- Replace use of `FootballField` with `FieldView` in the move scrollbar & main drill editor.
- Add a "print greyscale" option to the PDF export.
- Allow users to persistently customize color settings for Rank Panda.

### Before
![image](https://github.com/user-attachments/assets/3a981267-8bf1-4d4e-8342-a1be7ed530b6)

### After
![image](https://github.com/user-attachments/assets/8fc5c1e7-2dff-4e1a-9b19-4c53c630e04a)

vdye committed:
Point.java: inherit from Point2D

The custom 'Point' class we use is extremely similar to the 'Point2D.Float'
class. Let's inherit from that class so we can use Point2D methods and
hash/equality operators more easily.

Also, since Point2D.Float already includes 'getX()' and 'getY()' methods
(with the wrong return type), replace our methods of the same name with
'X()' and 'Y()'.

vdye committed:
views: add FieldStyle class

Introduce a new view(ish) class 'FieldStyle' that captures the view styling
information for rendered drill (colors, relative sizing, etc.). In later
commits, we will use instances of this class to represent that information
in the drill view classes themselves.

vdye committed:
Point.java: add math helpers

Add helper methods to our 'Point' class including:

- 'normalize' (get the unit vector that points to the Point instance)
- 'orthogonal' (get the vector orthogonal - but identical in magnitude - the
  vector pointing at the Point instance)
- 'multiply' (multiply the Point's X & Y values by a constant)
- 'add' (add the X & Y values, respectively, of two points)
- 'subtract' (subtract the X & Y values, respectively, of two points)

Most of these functions approximate vector math, which is useful for getting
the directionality of ranks. In later commits, this will greatly simplify
our drill rendering code.

Include function documentation for each new function, as well as adding it
for the existing methods 'distance' and 'interpolate'.

vdye committed:
RankPosition.java: add control point getter

Add the method 'getCurveControlPoint' to 'RankPosition' to get the linear
midpoint of the start and end of the rank, which represents the control
point for a curved rank. This will simplify some of the rank rendering code
in future commits.

vdye committed:
PdfImage.java: replace FootballField with FieldView

Create a new FieldView abstract class that represents a generic drill
FieldView, with the rendering functions simplified from their equivalents in
FootballField by using mathematical helper functions in Point and
transforming the coordinate system to feet before performing the actual
render. Update PdfImage to inherit from FieldView to use the new rendering
code.

The rewritten rendering code adjusts some sizing of things like the
arrowhead on ranks to ideally be slightly more visually appealing and/or
readable. For the most part, though, the drill looks the same as the
FootballField rendering. The main benefit is that the new code allows for
much easier customization based on user preferences (e.g. greyscale coloring
for the PDF).

In future commits, we will also replace the move thumbnails and main drill
editor with classes extending FieldView.

vdye committed:
PDFGenerator.java: print higher-resolution drill

Change the resolution of the drill image from 72 DPI to 300 DPI to eliminate
jagged rendering artifacts from the drill in the PDF.

vdye committed:
PDFGenerator.java: include sidelines in height

Scale the 'drillHeight' variable based on the total height of the football
field, rather than excluding the sidelines. For now, that doesn't make a
difference in the PDF rendering (since the FootballField class uses its own
measurements), but this will matter in future commits when we render based
on the 'Field' instance dimensions.

vdye committed:
FieldView.java: enable rendering quality settings

Enable antialiasing, bilinear interpolation, and "quality" rendering when
painting the drill. This will help avoid artifacts in the drill image.

vdye committed:
FieldView.java: add comment about rank offset error

Somehow, due to either the rank positioning or grid snapping code, all rank
positions are 0.6ft (0.2yd) further back on the field than they should be.
This is visible in the .pnd files, so it's not just a rendering error.

Instead of adding the offset to the render (which could, in the future,
introduce error itself if we fix the rank position issue), add a comment
describing the issue in detail. Separately, we can fix the snapping issue
and, ideally, add versioning to the .pnd file so we can adjust the affected
files automatically when saved in the future.

dmk257 commented on src/main/java/org/bigredbands/mb/views/FieldStyle.java:
```diff
@@ -0,0 +1,92 @@
+    // ****************************
+    // ***** Field properties *****
+    // ****************************
+    private float fieldNumberSize;
```
I know we don't have an official style guide for this project, but general Java style (and what is used elsewhere in the project) have getters named `getX()` and setters named `setY()`.  Can you update the getters and setters in this file to match?

vdye replied:
Ahh, you're right, I just hate how verbose Java is and went with a more C#-esque style here to get around it 😅. But consistency is more important than my hangups with Java, so I'll update accordingly.

dmk257 replied:
Yeah, I feel that.  Modern C++ style also tends to drop the "gets" and "sets" nowadays.

dmk257 commented on src/main/java/org/bigredbands/mb/views/FieldView.java:
```diff
@@ -0,0 +1,325 @@
+
+    @Override
+    public void paintComponent(Graphics g) {
+        super.paintComponent(g);
```
Should this happen after we set the rendering details below?

vdye replied:
I don't want to side-effect the `super` call (in this case, I think just "standard `JPanel`") with the rendering settings here, and instead only to apply them to the drill panel and any subcomponents (the field, ranks, etc.).

dmk257 replied:
Oh, I misread what this method was doing.  Feel free to ignore this comment.

dmk257 commented on src/main/java/org/bigredbands/mb/models/Point.java:
```diff
@@ -23,6 +23,26 @@ public Point interpolate(Point b, float t) {
+        return this.multiply(1.0f / (float) this.distance(0, 0));
+    }
+
+    public Point orthogonal() {
```
Can we rename this to something like `getOrthogonalVector` or something similar that clarifies this is meant to be used when a point represents a vector?

If this represents a 2D point (as it does when representing the start and end of a RankPosition), orthogonal doesn't make sense.

vdye replied:
I modeled these changes after the [`javafx.geometry.Point2D`](https://docs.oracle.com/javase/8/javafx/api/javafx/geometry/Point2D.html) class, which represents a point _or_ vector (as opposed to the AWT [`Point2D`](https://docs.oracle.com/javase%2F8%2Fdocs%2Fapi%2F%2F/java/awt/geom/Point2D.html), which only represents a point). The JavaFX one doesn't have an equivalent `orthogonal()` method, but it does have `normalize()` + a bunch of other operations that only apply to vectors (`dotProduct()`, `crossProduct()`, etc.).

So while I'd like to keep the names (to introduce slightly less churn to eventually cut over completely to JavaFX, although that's a huge refactor in and of itself), I can make the "point or vector" purpose of `Point` clearer with class & function documentation.

dmk257 replied:
That works for me.  As a case study, I was confused reading them for a while for how a 2D point could have these operations until I got further into the code, so I think having the additional clarity would help.

dmk257 commented on src/main/java/org/bigredbands/mb/models/Point.java:
```diff
@@ -23,6 +23,26 @@ public Point interpolate(Point b, float t) {
         return new Point(this.x*(1-t) + b.x*t, this.y*(1-t) + b.y*t);
     }
 
+    public Point normalize() {
```
Same comment about renaming.  Something like `normalizeVector`?

dmk257 commented on src/main/java/org/bigredbands/mb/views/PdfImage.java:
```diff
@@ -1,60 +1,21 @@
-    public float getScaleFactor(){
-        return this.scaleFactor;
+    @Override 
+    public HashMap<String, RankPosition> getRankPositions() {
```
Why not define this in the FieldView?

vdye replied:
Eventually, I plan to use this for the move thumbnails and the main editor as well. For `PdfImage`, getting the ranks is easy (they're just passed in from the `PDFGenerator`), but for the move thumbnail we'll need to call the controller to get the ranks for a specific move. The editor will need to get the ranks for the _current_ move, and possibly also deal with things like selected or transient ranks. So I went with the abstract method to allow each individual version of the view to define how it gets ranks.

dmk257 commented on src/main/java/org/bigredbands/mb/controllers/PDFGenerator.java:
```diff
@@ -116,8 +116,8 @@ public void createPDF(DrillInfo drillInfo, File file) throws IOException {
         // Units are in pixels
-        float imageWidth = drillWidth;
-        float imageHeight = drillHeight;
+        float imageWidth = drillWidth * (300f / 72f);
```
Let's add a comment here about why this multiplication is happening, and why the values are set to 300 and 72 respectively.

dmk257 reviewed (commented):
Generally LGTM.  Primarily a few small comments about renaming to make it easier to understand in the future.

vdye commented:
I think I've addressed all of the comments (added function documentation to `Point`, changed getters/setters in `FieldStyle` to begin with `get` and `set` respectively, added a comment explaining pixel scaling from 72dpi -> 300dpi). Here's the range-diff of the changes:

<details>
<summary>Range-diff vs. previous version</summary>

```diff
 1:  5b9302a =  1:  46d334c PDFGenerator.java: include sidelines in height
 2:  581c3e3 !  2:  7bdb41c Point.java: inherit from Point2D
    @@ src/main/java/org/bigredbands/mb/models/Point.java
     -    private Float x;
     -    private Float y;
     -
    ++/**
    ++ * The Point class represents a point or vector in two dimensional space,
    ++ * similar to the javafx.geometry.Point2D class. The X and Y coordinates of each
    ++ * point are stored and accessed as single-precision floating point values.
    ++ */
     +public class Point extends Point2D.Float {
          public Point(float x, float y){
     -        this.x = x;
 3:  8c6d569 !  3:  f7317fe Point.java: add math helpers
    @@ Commit message
         the directionality of ranks. In later commits, this will greatly simplify
         our drill rendering code.
     
    +    Include function documentation for each new function, as well as adding it
    +    for the existing methods 'distance' and 'interpolate'.
    +
      ## src/main/java/org/bigredbands/mb/models/Point.java ##
     @@ src/main/java/org/bigredbands/mb/models/Point.java: public class Point extends Point2D.Float {
    +         return this.y;
    +     }
    + 
    ++    /**
    ++     * Interpolate between this instance and a given point using a scalar value
    ++     * `t`.
    ++     *
    ++     * @param b The point to interpolate with this.
    ++     * @param t The relative weight of each point, where this point is weighted
    ++     *          with a factor `t` and `b` is weighted with `(1 - t)`.
    ++     * @return The interpolated Point.
    ++     */
    +     public Point interpolate(Point b, float t) {
              return new Point(this.x*(1-t) + b.x*t, this.y*(1-t) + b.y*t);
          }
      
    ++    /**
    ++     * Normalizes the relative magnitude vector represented by this instance.
    ++     *
    ++     * @return A point representing the unit vector of this point.
    ++     */
     +    public Point normalize() {
     +        return this.multiply(1.0f / (float) this.distance(0, 0));
     +    }
     +
    ++    /**
    ++     * Get the vector orthogonal to this vector, equivalent to rotating the
    ++     * point about origin counter-clockwise.
    ++     *
    ++     * @return The vector orthogonal and equal in magnitude to this.
    ++     */
     +    public Point orthogonal() {
     +        return new Point(-this.y, this.x);
     +    }
     +
    ++    /**
    ++     * Returns a point with the coordinates of this point multiplied by the
    ++     * specified scalar factor.
    ++     *
    ++     * @param value The scalar multiplier.
    ++     * @return The product of this point and the scalar `value`.
    ++     */
     +    public Point multiply(float value) {
     +        return new Point(this.x * value, this.y * value);
     +    }
     +
    ++    /**
    ++     * Returns a point with the coordinates of the specified point added to the
    ++     * coordinates of this point.
    ++     *
    ++     * @param other The point to add to this point
    ++     * @return The sum of this and `other`
    ++     */
     +    public Point add(Point other) {
     +        return new Point(this.x + (float) other.getX(), this.y + (float) other.getY());
     +    }
     +
    ++    /**
    ++     * Returns a point with the coordinates of the specified point subtracted
    ++     * from the coordinates of this point.
    ++     *
    ++     * @param other The point to subtract from this point
    ++     * @return The difference between this and `other`.
    ++     */
     +    public Point subtract(Point other) {
     +        return this.add(other.multiply(-1.0f));
     +    }
    ++
    ++    /**
    ++     * Computes the distance between this point and point `(x, y)`.
    ++     *
    ++     * @param x The X coordinate of the other point.
    ++     * @param y The Y coordinate of the other point.
    ++     * @return The distance between this and the point at the given coordinate.
    ++     */
    ++    public double distance(double x, double y) {
    ++        return Math.sqrt((this.x - x) * (this.x - x) +
    ++                         (this.y - y) * (this.y - y));
    ++    }
     +
          public void setPoint(float x, float y){
              this.x = x;
              this.y = y;
    +@@ src/main/java/org/bigredbands/mb/models/Point.java: public class Point extends Point2D.Float {
    +         return "Point [x=" + x + ", y=" + y + "]";
    +     }
    + 
    +-    public double distance(double x, double y){
    +-        return Math.sqrt((this.x-x)*(this.x-x)+(this.y-y)*(this.y-y));
    +-    }
    +-
    +     public Element convertToXML(Document document, String pointName) {
    +         //add the point parent tag
    +         Element pointTag = document.createElement(pointName);
 4:  5a13dee =  4:  d632e45 RankPosition.java: add control point getter
 5:  1f80be2 !  5:  d9d2f3f views: add FieldStyle class
    @@ src/main/java/org/bigredbands/mb/views/FieldStyle.java (new)
     +    // ***** Field properties *****
     +    // ****************************
     +    private float fieldNumberSize;
    -+    public float FieldNumberSize() { return fieldNumberSize; }
    -+    public FieldStyle FieldNumberSize(float value) { fieldNumberSize = value; return this; }
    ++    public float getFieldNumberSize() { return fieldNumberSize; }
    ++    public FieldStyle setFieldNumberSize(float value) { fieldNumberSize = value; return this; }
     +
     +    private float hashWidth;
    -+    public float HashWidth() { return hashWidth; }
    -+    public FieldStyle HashWidth(float value) { hashWidth = value; return this; }
    ++    public float getHashWidth() { return hashWidth; }
    ++    public FieldStyle setHashWidth(float value) { hashWidth = value; return this; }
     +
     +    private float hashPeriod;
    -+    public float HashPeriod() { return hashPeriod; }
    -+    public FieldStyle HashPeriod(float value) { hashPeriod = value; return this; }
    ++    public float getHashPeriod() { return hashPeriod; }
    ++    public FieldStyle setHashPeriod(float value) { hashPeriod = value; return this; }
     +
     +    private float majorIncrementWidth;
    -+    public float MajorIncrementWidth() { return majorIncrementWidth; }
    -+    public FieldStyle MajorIncrementWidth(float value) { majorIncrementWidth = value; return this; }
    ++    public float getMajorIncrementWidth() { return majorIncrementWidth; }
    ++    public FieldStyle setMajorIncrementWidth(float value) { majorIncrementWidth = value; return this; }
     +
     +    private float minorIncrementWidth;
    -+    public float MinorIncrementWidth() { return minorIncrementWidth; }
    -+    public FieldStyle MinorIncrementWidth(float value) { minorIncrementWidth = value; return this; }
    ++    public float getMinorIncrementWidth() { return minorIncrementWidth; }
    ++    public FieldStyle setMinorIncrementWidth(float value) { minorIncrementWidth = value; return this; }
     +
     +    private float gridWidth;
    -+    public float GridWidth() { return gridWidth; }
    -+    public FieldStyle GridWidth(float value) { gridWidth = value; return this; }
    ++    public float getGridWidth() { return gridWidth; }
    ++    public FieldStyle setGridWidth(float value) { gridWidth = value; return this; }
     +
     +    private Color backgroundColor;
    -+    public Color BackgroundColor() { return backgroundColor; }
    -+    public FieldStyle BackgroundColor(Color value) { backgroundColor = value; return this; }
    ++    public Color getBackgroundColor() { return backgroundColor; }
    ++    public FieldStyle setBackgroundColor(Color value) { backgroundColor = value; return this; }
     +
     +    private Color fieldColor;
    -+    public Color FieldColor() { return fieldColor; }
    -+    public FieldStyle FieldColor(Color value) { fieldColor = value; return this; }
    ++    public Color getFieldColor() { return fieldColor; }
    ++    public FieldStyle setFieldColor(Color value) { fieldColor = value; return this; }
     +
     +    // ***************************
     +    // ***** Rank properties *****
     +    // ***************************
     +    private float arrowWidth;
    -+    public float ArrowWidth() { return arrowWidth; }
    -+    public FieldStyle ArrowWidth(float value) { arrowWidth = value; return this; }
    ++    public float getArrowWidth() { return arrowWidth; }
    ++    public FieldStyle setArrowWidth(float value) { arrowWidth = value; return this; }
     +
     +    private float rankEndDiameter;
    -+    public float RankEndDiameter() { return rankEndDiameter; }
    -+    public FieldStyle RankEndDiameter(float value) { rankEndDiameter = value; return this; }
    ++    public float getRankEndDiameter() { return rankEndDiameter; }
    ++    public FieldStyle setRankEndDiameter(float value) { rankEndDiameter = value; return this; }
     +
     +    private float rankLabelSize;
    -+    public float RankLabelSize() { return rankLabelSize; }
    -+    public FieldStyle RankLabelSize(float value) { rankLabelSize = value; return this; }
    ++    public float getRankLabelSize() { return rankLabelSize; }
    ++    public FieldStyle setRankLabelSize(float value) { rankLabelSize = value; return this; }
     +
     +    private float rankStrokeWidth;
    -+    public float RankStrokeWidth() { return rankStrokeWidth; }
    -+    public FieldStyle RankStrokeWidth(float value) { rankStrokeWidth = value; return this; }
    ++    public float getRankStrokeWidth() { return rankStrokeWidth; }
    ++    public FieldStyle setRankStrokeWidth(float value) { rankStrokeWidth = value; return this; }
     +
     +    private Color rankColor;
    -+    public Color RankColor() { return rankColor; }
    -+    public FieldStyle RankColor(Color value) { rankColor = value; return this; }
    ++    public Color getRankColor() { return rankColor; }
    ++    public FieldStyle setRankColor(Color value) { rankColor = value; return this; }
     +
     +    private Color rankLabelColor;
    -+    public Color RankLabelColor() { return rankLabelColor; }
    -+    public FieldStyle RankLabelColor(Color value) { rankLabelColor = value; return this; }
    ++    public Color getRankLabelColor() { return rankLabelColor; }
    ++    public FieldStyle setRankLabelColor(Color value) { rankLabelColor = value; return this; }
     +
     +    private Color rankLabelBackground;
    -+    public Color RankLabelBackground() { return rankLabelBackground; }
    -+    public FieldStyle RankLabelBackground(Color value) { rankLabelBackground = value; return this; }
    ++    public Color getRankLabelBackground() { return rankLabelBackground; }
    ++    public FieldStyle setRankLabelBackground(Color value) { rankLabelBackground = value; return this; }
     +
     +    // All measurements are in *feet* to allow for uniform UI scaling
     +    public FieldStyle() {
 6:  ba25722 !  6:  8703f9f PdfImage.java: replace FootballField with FieldView
    @@ src/main/java/org/bigredbands/mb/views/FieldView.java (new)
     +
     +    private void drawField(Graphics2D g) {
     +        // Paint the canvas background
    -+        g.setColor(fieldStyle.BackgroundColor());
    ++        g.setColor(fieldStyle.getBackgroundColor());
     +        g.fill(g.getClipBounds());
     +
     +        Rectangle2D fieldRect = new Rectangle2D.Float(field.EndzoneWidth, field.SidelineWidth,
     +            field.Length, field.Height);
    -+        g.setColor(fieldStyle.FieldColor());
    ++        g.setColor(fieldStyle.getFieldColor());
     +        g.fill(fieldRect);
     +
     +        g.setColor(Color.BLACK);
    -+        g.setStroke(new BasicStroke(fieldStyle.MajorIncrementWidth(),
    ++        g.setStroke(new BasicStroke(fieldStyle.getMajorIncrementWidth(),
     +            BasicStroke.CAP_BUTT,
     +            BasicStroke.JOIN_MITER));
     +        g.draw(fieldRect);
    @@ src/main/java/org/bigredbands/mb/views/FieldView.java (new)
     +     */
     +    private void drawHashes(Graphics2D g) {
     +        BasicStroke dashed = new BasicStroke(
    -+            fieldStyle.HashWidth(),
    ++            fieldStyle.getHashWidth(),
     +            BasicStroke.CAP_BUTT,
     +            BasicStroke.JOIN_MITER,
     +            10.0f,
    -+            new float[]{ fieldStyle.HashPeriod() },
    ++            new float[]{ fieldStyle.getHashPeriod() },
     +            0.0f);
     +
     +        g.setStroke(dashed);
    @@ src/main/java/org/bigredbands/mb/views/FieldView.java (new)
     +        int increments = (int) (field.Length / field.Increment);
     +
     +        g.setColor(Color.BLACK);
    -+        g.setFont(new Font(Font.SANS_SERIF, 0, (int) fieldStyle.FieldNumberSize()));
    ++        g.setFont(new Font(Font.SANS_SERIF, 0, (int) fieldStyle.getFieldNumberSize()));
     +        FontMetrics metrics = g.getFontMetrics();
     +
     +        int markedIncrements = increments / field.MajorIncrementFrequency;
    @@ src/main/java/org/bigredbands/mb/views/FieldView.java (new)
     +        // Divide the field horizontally into increments & draw each one
     +        int increments = (int) (field.Length / field.Increment);
     +        for (int i = 0; i < increments + 1; i++) {
    -+            float strokeWidth = fieldStyle.GridWidth();
    ++            float strokeWidth = fieldStyle.getGridWidth();
     +            if (i % field.MajorIncrementFrequency == 0) {
    -+                strokeWidth = fieldStyle.MajorIncrementWidth();
    ++                strokeWidth = fieldStyle.getMajorIncrementWidth();
     +            } else if (i % field.MinorIncrementFrequency == 0) {
    -+                strokeWidth = fieldStyle.MinorIncrementWidth();
    ++                strokeWidth = fieldStyle.getMinorIncrementWidth();
     +            }
     +            if (strokeWidth == 0) {
     +                continue;
    @@ src/main/java/org/bigredbands/mb/views/FieldView.java (new)
     +
     +        increments = (int) (field.Height / field.Increment);
     +        for (int i = 0; i < increments + 1; i++) {
    -+            float strokeWidth = fieldStyle.GridWidth();
    ++            float strokeWidth = fieldStyle.getGridWidth();
     +
     +            // Unlike horizontal increments, do not include major increment style
     +            if (i % field.MinorIncrementFrequency == 0) {
    -+                strokeWidth = fieldStyle.MinorIncrementWidth();
    ++                strokeWidth = fieldStyle.getMinorIncrementWidth();
     +            }
     +            if (strokeWidth == 0) {
     +                continue;
    @@ src/main/java/org/bigredbands/mb/views/FieldView.java (new)
     +    private void drawArrowhead(Graphics2D g, Point startPx, Point direction) {
     +        Path2D arrowhead = new Path2D.Float();
     +        Point dirNorm = direction.normalize();
    -+        float arrowHeightPx = (float) Math.sin(Math.toRadians(60)) * fieldStyle.ArrowWidth();
    -+        float arrowBasePx = fieldStyle.ArrowWidth();
    ++        float arrowHeightPx = (float) Math.sin(Math.toRadians(60)) * fieldStyle.getArrowWidth();
    ++        float arrowBasePx = fieldStyle.getArrowWidth();
     +
     +        // The arrowhead point
     +        Point vertex = startPx.add(dirNorm.multiply(arrowHeightPx / 2.0f));
    @@ src/main/java/org/bigredbands/mb/views/FieldView.java (new)
     +
     +        // Draw the rank arrow
     +        // Draw the line
    -+        g.setStroke(new BasicStroke(fieldStyle.RankStrokeWidth(),
    ++        g.setStroke(new BasicStroke(fieldStyle.getRankStrokeWidth(),
     +            BasicStroke.CAP_BUTT, BasicStroke.JOIN_MITER));
    -+        g.setColor(fieldStyle.RankColor());
    ++        g.setColor(fieldStyle.getRankColor());
     +
     +        // TODO: convert rank scale to feet (avoid the multiply())
     +        Point start = rank.getFront().multiply(3.0f).add(fieldOffset);
    @@ src/main/java/org/bigredbands/mb/views/FieldView.java (new)
     +        }
     +
     +        // Draw endpoints
    -+        g.fill(new Arc2D.Float((end.X() - 0.5f * fieldStyle.RankEndDiameter()),
    -+            (end.Y() - 0.5f * fieldStyle.RankEndDiameter()),
    -+            fieldStyle.RankEndDiameter(),
    -+            fieldStyle.RankEndDiameter(),
    ++        g.fill(new Arc2D.Float((end.X() - 0.5f * fieldStyle.getRankEndDiameter()),
    ++            (end.Y() - 0.5f * fieldStyle.getRankEndDiameter()),
    ++            fieldStyle.getRankEndDiameter(),
    ++            fieldStyle.getRankEndDiameter(),
     +            0.0f, 360.0f,
     +            Arc2D.CHORD));
     +        drawArrowhead(g, start, arrowDir);
     +
     +        // Draw the rank label
    -+        Font rankFont = new Font(Font.SANS_SERIF, Font.BOLD, (int) fieldStyle.RankLabelSize());
    ++        Font rankFont = new Font(Font.SANS_SERIF, Font.BOLD, (int) fieldStyle.getRankLabelSize());
     +        GlyphVector rankLabel = rankFont.createGlyphVector(g.getFontRenderContext(), rankName);
     +        Shape rankShape = rankLabel.getOutline(midpoint.X(), midpoint.Y());
     +
    -+        g.setStroke(new BasicStroke(0.1f * fieldStyle.RankLabelSize()));
    -+        g.setColor(fieldStyle.RankLabelBackground());
    ++        g.setStroke(new BasicStroke(0.1f * fieldStyle.getRankLabelSize()));
    ++        g.setColor(fieldStyle.getRankLabelBackground());
     +        g.draw(rankShape);
    -+        g.setColor(fieldStyle.RankLabelColor());
    ++        g.setColor(fieldStyle.getRankLabelColor());
     +        g.fill(rankShape);
     +    }
     +
 7:  5fe966e =  7:  b90180f FieldView.java: add comment about rank offset error
 8:  7c8a617 =  8:  0139876 FieldView.java: enable rendering quality settings
 9:  14d8a44 !  9:  4b71411 PDFGenerator.java: print higher-resolution drill
    @@ Commit message
     
      ## src/main/java/org/bigredbands/mb/controllers/PDFGenerator.java ##
     @@ src/main/java/org/bigredbands/mb/controllers/PDFGenerator.java: public class PDFGenerator {
    +         float lineSpacing = 1.5f;
      
              // Drill image dimensions
    -         // Units are in pixels
    +-        // Units are in pixels
     -        float imageWidth = drillWidth;
     -        float imageHeight = drillHeight;
    ++        // Units are in pixels, scaled from 72dpi (the units of drillWidth and
    ++        // drillHeight) to 300dpi.
     +        float imageWidth = drillWidth * (300f / 72f);
     +        float imageHeight = drillHeight * (300f / 72f);
              Dimension dim = new Dimension((int) imageWidth, (int) imageHeight);
```

</details>
